wpilibj/src/main/java/edu/wpi/first/wpilibj/experimental/controller/PIDController.java

```diff
@@ -0,0 +1,581 @@
+    }
+  }
+
+  /**
+   * Get the Integral coefficient.
+   *
+   * @return integral coefficient
+   */
+  public double getI() {
+    m_thisMutex.lock();
+    try {
+      return m_Ki;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Get the Differential coefficient.
+   *
+   * @return differential coefficient
+   */
+  public double getD() {
+    m_thisMutex.lock();
+    try {
+      return m_Kd;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Returns the period of this controller.
+   *
+   * @return the period of the controller.
+   */
+  public double getPeriod() {
+    return m_period;
+  }
+
+  /**
+   * Return the current controller output.
+   *
+   * <p>This is always centered around zero and constrained to the min and max outputs.
+   *
+   * @return The latest calculated output.
+   */
+  public double getOutput() {
+    m_thisMutex.lock();
+    try {
+      return m_output;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Set the reference for the PIDController.
+   *
+   * @param reference the desired reference
+   */
+  public void setReference(double reference) {
+    m_thisMutex.lock();
+    try {
+      if (m_maximumInput > m_minimumInput) {
+        m_reference = clamp(reference, m_minimumInput, m_maximumInput);
+      } else {
+        m_reference = reference;
+      }
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Returns the current reference of the PIDController.
+   *
+   * @return the current reference
+   */
+  public double getReference() {
+    m_thisMutex.lock();
+    try {
+      return m_reference;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Return true if the error is within the percentage of the total input range,
+   * determined by SetTolerance. This asssumes that the maximum and minimum
+   * input were set using SetInput.
+   *
+   * <p>Currently this just reports on target as the actual value passes through the setpoint.
+   * Ideally it should be based on being within the tolerance for some period of time.
+   *
+   * <p>This will return false until at least one input value has been computed.
+   *
+   * @return Whether the error is within the acceptable bounds.
+   */
+  public boolean atReference() {
+    return atReference(m_tolerance, m_deltaTolerance, m_toleranceType);
+  }
+
+  /**
+   * Return true if the error and change in error are below the specified tolerances.
+   *
+   * @param tolerance The maximum allowable error.
+   * @param deltaTolerance The maximum allowable change in error from the previous iteration.
+   * @param toleranceType Whether the given tolerance values are absolute, or percentages of the
+   *                      total input range.
+   * @return Whether the error is within the acceptable bounds.
+   */
+  public boolean atReference(double tolerance, double deltaTolerance, Tolerance toleranceType) {
+    double error = getError();
```

Comment:
Why not get the error inside of the lock?

Reply:
Because getError() locks the mutex as well, and we want to avoid recursive locking for performance reasons. It might not matter in Java because all locks are recursive anyway, but it does matter in C++, so we do the same here.